File: lib/csharp/microsoft.bot.builder.skills/Microsoft.Bot.Builder.Skills/Auth/JwtClaimAuthProvider.cs
using Microsoft.AspNetCore.Http;
using System;
using System.Linq;
using System.Security.Claims;

namespace Microsoft.Bot.Builder.Skills.Auth
{
    public class JwtClaimAuthProvider : ISkillAuthProvider
    {
        private readonly ISkillWhitelist _skillWhitelist;

        public JwtClaimAuthProvider(ISkillWhitelist skillWhitelist)
        {
            _skillWhitelist = skillWhitelist;
        }

        public bool Authenticate(HttpContext httpContext)
        {
            if (httpContext == null)
            {
                throw new ArgumentNullException(nameof(HttpContext));
            }

            // if not provided a list, ignore this authentication operation
            if (_skillWhitelist == null || _skillWhitelist.SkillWhiteList == null || _skillWhitelist.SkillWhiteList.Count() == 0)
            {
                return true;
            }

            var identity = httpContext.User.Identity;
            if (identity is ClaimsIdentity claims)
            {
                var version = claims.Claims.FirstOrDefault(c => c.Type.Equals("ver", StringComparison.InvariantCultureIgnoreCase));
                var appId = string.Empty;
                var appIdClaimType = string.Empty;
                if (version.Value.Equals("1.0"))
                {
                    appIdClaimType = "appid";
                }
                else if (version.Value.Equals("2.0"))
                {
                    appIdClaimType = "azp";
                }

                var appIdClaim = claims.Claims.FirstOrDefault(c => c.Type.Equals(appIdClaimType, StringComparison.InvariantCulture));
                if (appIdClaim != null)
                {
                    appId = appIdClaim.Value;
                }

                if (_skillWhitelist.SkillWhiteList.Contains(appId))
                {
                    return true;
                }
            }

            return false;
        }
    }
}

File: lib/typescript/botbuilder-skills/src/auth/jwtClaimAuthProvider.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { ClaimsIdentity, ICredentialProvider, JwtTokenValidation } from 'botframework-connector';
import { Activity } from 'botframework-schema';
import { ISkillAuthProvider } from './skillAuthProvider';
import { ISkillWhitelist } from './skillWhitelist';

export class JwtClaimAuthProvider implements ISkillAuthProvider {
    private readonly skillWhitelist: ISkillWhitelist;
    private readonly credentialProvider: ICredentialProvider;

    constructor(skillWhitelist: ISkillWhitelist, credentialProvider: ICredentialProvider) {
        this.skillWhitelist = skillWhitelist;
        this.credentialProvider = credentialProvider;
    }

    public async authenticate(authHeader: string, activity: Activity, channelService?: string): Promise<boolean> {
        // if not provided a list, ignore this authentication operation
        if (this.skillWhitelist === undefined
            || this.skillWhitelist.skillWhiteList === undefined
            || this.skillWhitelist.skillWhiteList.length === 0) {
            return true;
        }

        try {
            const claims: ClaimsIdentity = await JwtTokenValidation.authenticateRequest(
                activity,
                authHeader,
                this.credentialProvider,
                channelService || '');
            const version: string = claims.getClaimValue('ver') || '';
            let appId: string = '';
            let appIdClaimType: string = '';

            if (version === '1.0') {
                appIdClaimType = 'appid';
            } else if (version === '2.0') {
                appIdClaimType = 'azip';
            }

            const appIdClaim: string = claims.getClaimValue(appIdClaimType) || '';
            if (appIdClaim) {
                appId = appIdClaim;
            }

            return this.skillWhitelist.skillWhiteList.includes(appId);
        } catch {
            return false;
        }
    }
}
